### PR TITLE
[fix](be) Fix creating hdfs connection coredump within bthread

### LINKED ISSUE #42610: [Bug] doris::io::HdfsHandlerCache::get_connection coredump
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master (cloud mode)
de93c26b86821fbaf845812423ecc8ff880b934a

### What's Wrong?
```
[WARNING!] /sys/kernel/mm/transparent_hugepage/enabled: [always] madvise never, Doris not recommend turning on THP, which may cause the BE process to use more me
mory and cannot be freed in time. Turn off THP: `echo madvise | sudo tee /sys/kernel/mm/transparent_hugepage/enabled`
start BE in cloud mode, cloud_unique_id: compute_node_1, meta_service_endpoint: 161.90.101.1:5000
Call to AttachCurrentThread failed with error: -1
getJNIEnv: getGlobalJNIEnv failed
*** Query id: 0-0 ***
*** is nereids: 0 ***
*** tablet id: 0 ***
*** Aborted at 1729654413 (unix time) try "date -d @1729654413" if you are using GNU date ***
*** Current BE git commitID: f65223fdb3 ***
*** SIGSEGV address not mapped to object (@0x10) received by PID 761 (TID 2947 OR 0x7efc4aecc700) from PID 16; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/common/signal_hand
ler.h:421
 1# PosixSignals::chained_handler(int, siginfo*, void*) [clone .part.0] in /usr/local/openjdk-17/lib/server/libjvm.so
 2# JVM_handle_linux_signal in /usr/local/openjdk-17/lib/server/libjvm.so
 3# 0x00007F00C11D2D60 in /lib/x86_64-linux-gnu/libc.so.6
 4# getLastTLSExceptionRootCause at /mnt/disk1/zhanglei/work/repo/incubator-doris/thirdparty/src/doris-thirdparty-hadoop-3.3.6.3-for-doris/hadoop-hdfs-project/ha
doop-hdfs-native-client/src/main/native/libhdfs/jni_helper.c:866
 5# doris::io::hdfs_error[abi:cxx11]() in /opt/apache-doris/be/lib/doris_be
 6# doris::io::HdfsHandlerCache::get_connection(doris::THdfsParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&
, std::shared_ptr<doris::io::HdfsHandler>*) at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/io/hdfs_util.cpp:134
 7# doris::io::HdfsFileSystem::init() at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/io/fs/hdfs_file_system.cpp:91
 8# doris::io::HdfsFileSystem::create(doris::THdfsParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::b
asic_string<char, std::char_traits<char>, std::allocator<char> >, doris::RuntimeProfile*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator
<char> >) at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/io/fs/hdfs_file_system.cpp:73
 9# doris::VaultCreateFSVisitor::operator()(doris::cloud::HdfsVaultInfo const&) const at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/cloud/cloud_storage
_engine.cpp:120
10# std::__detail::__variant::__gen_vtable_impl<std::__detail::__variant::_Multi_array<std::__detail::__variant::__deduce_visit_result<doris::Status> (*)(doris::
VaultCreateFSVisitor&&, std::variant<doris::S3Conf, doris::cloud::HdfsVaultInfo>&)>, std::integer_sequence<unsigned long, 1ul> >::__visit_invoke(doris::VaultCreateFSVisitor&&, std::variant<doris::S3Conf, doris::cloud::HdfsVaultInfo>&) at /mnt/disk1/zhanglei/work/app/ldb_toolchain-16/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/variant:1013
11# doris::CloudStorageEngine::sync_storage_vault() at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/cloud/cloud_storage_engine.cpp:341
12# doris::CloudStorageEngine::get_storage_resource(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/cloud/cloud_storage_engine.h:82
13# doris::CloudRowsetBuilder::init() at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/cloud/cloud_rowset_builder.cpp:73
14# doris::BaseDeltaWriter::init() at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/olap/delta_writer.cpp:106
15# std::_Function_handler<doris::Status (), doris::CloudDeltaWriter::batch_init(std::vector<doris::CloudDeltaWriter*, std::allocator<doris::CloudDeltaWriter*> >
)::$_0>::_M_invoke(std::_Any_data const&) at /mnt/disk1/zhanglei/work/app/ldb_toolchain-16/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std
_function.h:291
16# std::_Function_handler<void (), doris::cloud::bthread_fork_join(std::vector<std::function<doris::Status ()>, std::allocator<std::function<doris::Status ()> >
 > const&, int)::$_1>::_M_invoke(std::_Any_data const&) at /mnt/disk1/zhanglei/work/app/ldb_toolchain-16/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c
++/11/bits/std_function.h:291
17# doris::cloud::bthread_fork_join(std::vector<std::function<doris::Status ()>, std::allocator<std::function<doris::Status ()> > > const&, int)::$_0::__invoke(v
oid*) at /mnt/disk1/zhanglei/work/repo/incubator-doris/be/src/cloud/cloud_meta_mgr.cpp:76
18# bthread::TaskGroup::task_runner(long) in /opt/apache-doris/be/lib/doris_be
19# bthread_make_fcontext in /opt/apache-doris/be/lib/doris_be 
```

### What You Expected?

doris_be work well

### How to Reproduce?

1. deploy doris with cloud mode
2. create a hdfs vault
3. create a table using the hdfs vault
4. insert into a row data into the data and be will coredump

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by swjtu-zhanglei**
* https://brpc.apache.org/docs/server/basics/
* According to the brpc doc, `JNI code checks stack layout and cannot be run in bthreads` so create a pthread for creating hdfs connection if necessary.

## Proposed changes

Issue Number: close #42610

### CODE REVIEW

**Comment by SWJTU-ZhangLei** [comment]
run buildall

File: be/src/io/hdfs_util.cpp [hunk 1]
```diff
@@ -41,6 +44,54 @@ Status create_hdfs_fs(const THdfsParams& hdfs_params, const std::string& fs_name
     return Status::OK();
 }
 
+Status create_hdfs_fs_async(const THdfsParams& hdfs_params, const std::string& fs_name,
+                            hdfsFS* fs) {
+    std::promise<std::pair<Status, hdfsFS>> prom;
+    std::future<std::pair<Status, hdfsFS>> fut = prom.get_future();
+
+    std::thread t(
+            [&](std::promise<std::pair<Status, hdfsFS>> prom) {
```
Comment (gavinchou) [inline be/src/io/hdfs_util.cpp:53]: why not just call `create_hdfs_fs_sync()` here?

Comment (SWJTU-ZhangLei) [reply be/src/io/hdfs_util.cpp:53]: > why not just call `create_hdfs_fs_sync()` here?

for returning multi value

**Comment by SWJTU-ZhangLei** [comment]
run buildall

**Comment by SWJTU-ZhangLei** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

File: be/src/io/hdfs_util.cpp [hunk 2]
```diff
@@ -41,6 +43,22 @@ Status create_hdfs_fs(const THdfsParams& hdfs_params, const std::string& fs_name
...
+    if (is_bthread()) {
+        Status st;
+        std::thread t([&] { st = _create_hdfs_fs(hdfs_params, fs_name, fs); });
+        if (t.joinable()) {
+            t.join();
+        }
+        return st;
+    }
+    return _create_hdfs_fs(hdfs_params, fs_name, fs);
+}
```
Comment (gavinchou) [inline be/src/io/hdfs_util.cpp:60]:
```suggestion
// https://brpc.apache.org/docs/server/basics/
// According to the brpc doc, JNI code checks stack layout and cannot be run in
// bthreads so create a pthread for creating hdfs connection if necessary.
Status create_hdfs_fs(const THdfsParams& hdfs_params, const std::string& fs_name, hdfsFS* fs) {
    if (bthread_self() != 0) { // running in bthread
        Status st;
        std::thread t([&] { st = _create_hdfs_fs(hdfs_params, fs_name, fs); });
        if (t.joinable()) {
            t.join();
        }
        return st;
    }
    return _create_hdfs_fs(hdfs_params, fs_name, fs);
}
```

**Comment by gavinchou** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Review by bobhan1 — APPROVED** [review]

File: be/src/io/hdfs_util.cpp [hunk 3]
```diff
@@ -41,6 +43,21 @@ Status create_hdfs_fs(const THdfsParams& hdfs_params, const std::string& fs_name
...
+    if (bthread_self() != 0) { // running in bthread
+        Status st;
+        std::thread t([&] { st = _create_hdfs_fs(hdfs_params, fs_name, fs); });
+        if (t.joinable()) {
+            t.join();
+        }
+        return st;
+    }
+    return _create_hdfs_fs(hdfs_params, fs_name, fs);
+}
```
Comment (gavinchou) [inline be/src/io/hdfs_util.cpp:59]:
```suggestion
// https://brpc.apache.org/docs/server/basics/
// According to the brpc doc, JNI code checks stack layout and cannot be run in
// bthreads so create a pthread for creating hdfs connection if necessary.
Status create_hdfs_fs(const THdfsParams& hdfs_params, const std::string& fs_name, hdfsFS* fs) {
    if (bthread_self() != 0) { // running in bthread
        Status st;
        auto btx = butex_create();
        std::atomic_bool done = false;
        std::thread t([&] {
            st = _create_hdfs_fs(hdfs_params, fs_name, fs);
            done = true;
            butex_wake(btx);
        });
        if (!done) butex_wait(btx);
        butex_destroy(btx);
        if (t.joinable()) {
            t.join();
        }
        return st;
    }
    return _create_hdfs_fs(hdfs_params, fs_name, fs);
}
```

**Comment by gavinchou** [comment]
run buildall

**Comment by gavinchou** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Pull request merged**